### PR TITLE
Firefox Bug Fix: Dynamic Sample Rate Detection for Audio Compatibility

### PR DESCRIPTION
**Pull request by mmtmn**
On Firefox, we get the [issue](https://github.com/openai/openai-realtime-console/issues/5):

Here is the explanation to this fix:

`Connecting AudioNodes from AudioContexts with different sample-rate is currently not supported.` 

- getSystemSampleRate(): This helper function will dynamically detect the system's default sample rate by creating a new AudioContext instance and querying its sampleRate property. This is important because different browsers and devices may have different default sample rates, and this will prevent sample rate mismatch issues.
- 
- Changes in WavStreamPlayer and WavRecorder: By using new AudioContext().sampleRate as the fallback, the application adapts to the system's default sample rate if none is explicitly provided. This is critical for maintaining compatibility across different browsers, particularly Firefox, which enforces stricter rules on sample rates.
- 

These are the changes needed to ensure your application works seamlessly with the system's default audio configuration without causing issues, particularly in Firefox.

### CODE REVIEW

**Comment by khorwood-openai** [comment]
For the Realtime API we need a 24,000 Hz sample rate. Do you think there's another solution to this problem? This is Firefox only, correct?

**Comment by Grant-CP** [comment]
@khorwood-openai From my brief testing this happens on Firefox and not Safari.

**Comment by mmtmn** [comment]
> For the Realtime API we need a 24,000 Hz sample rate. Do you think there's another solution to this problem? This is Firefox only, correct?

Ah, I see. I think there might be a way around this but I'm not sure. I'll report back in a few minutes.

**Comment by mmtmn** [comment]
Working on a resample Audio to 24,000 Hz solution to this pull request, should be updating this shortly

**Comment by mmtmn** [comment]
Ok, I think resampling should be working, just improving a bit on the resampling efficiency. Once I get the resampling efficient I'll update the PR.

**Comment by mmtmn** [comment]
@khorwood-openai I've added resampling to this pull request, it should be good to review now.

Please just keep in mind that the realtime endpoint is not available yet, so I think it works, but I can't be 100% sure of it as after connecting I get: `Could not connect to "wss://api.openai.com/v1/realtime"`

**Comment by samiy803** [comment]
@khorwood-openai any updates on this? This fixes issue #5

**Comment by butelo** [comment]
hello
is not working on my side, some import errors

`int16ArrayToAudioBuffer is not defined
processChunk@http://localhost:3002/static/js/bundle.js:60860:27
add16BitPCM@http://localhost:3002/static/js/bundle.js:60880:13
./src/pages/ConsolePage.tsx/ConsolePage/</<@http://localhost:3002/static/js/bundle.js:1026:25
dispatch@http://localhost:3002/static/js/bundle.js:59326:14
handlerWithDispatch@http://localhost:3002/static/js/bundle.js:58327:14
dispatch@http://localhost:3002/static/js/bundle.js:59326:14
receive@http://localhost:3002/static/js/bundle.js:57944:10
./src/lib/realtime-api-beta/lib/api.js/connect/<@http://localhost:3002/static/js/bundle.js:57845:14`

manually fixing the imports "kind of works" but the speech is broken, is too fast. same happens on other browsers. This happens on a Mac computer, didn't try other

**Comment by mmtmn** [comment]
> hello is not working on my side, some import errors
I'll find some time to check this out tmrw

**Comment by mmtmn** [comment]
ok, so, there were a missing exports and imports, I should update this PR soon

**Comment by mmtmn** [comment]
ok, so, this is working on firefox now, but it doesn't work as well on chrome for example. this is all the time I had for today to contribute. I'll check out whats going wrong on chrome on this pull request when I can, hopefully soon.

File: src/lib/wavtools/lib/wav_stream_player.js [hunk 1]
```diff
@@ -119,23 +121,17 @@ export class WavStreamPlayer {
...
-  
-      // Convert Int16Array to AudioBuffer
-      const audioBuffer = await int16ArrayToAudioBuffer(chunkBuffer, this.sampleRate);
-  
-      // Resample to 24,000 Hz
-      const resampledBuffer = await resampleAudioBuffer(audioBuffer, 24000);
-  
-      // Convert the resampled AudioBuffer back to Int16Array
+      const audioBuffer = await int16ArrayToAudioBuffer(chunkBuffer, apiSampleRate); // Use apiSampleRate directly
+      const resampledBuffer = await resampleAudioBuffer(audioBuffer, apiSampleRate);
```
Comment (BasedUser) [inline src/lib/wavtools/lib/wav_stream_player.js:128]: Shouldn't this be the audio context rate?
```ts
      const resampledBuffer = await resampleAudioBuffer(audioBuffer, this.samplerate);
```
On Firefox, *not* doing this results in a squeaky voice - but also stuttering. Then again, I'm a terrible tester.

Comment (mmtmn) [reply src/lib/wavtools/lib/wav_stream_player.js:128]: all good! I haven't had the time to fully finish this pull request

**Comment by mmtmn** [comment]
I'm closing this pull request for now, I don't have time to fully commit to making a high quality contribution for now, I'll revisit this in a few months in case this is still something needed.

note to future self:

- it seems to be working in firefox but I'd review the logic again
- it breaks in anything that is running on chromium engine, this pull request is not good enough if it fixes one thing and breaks everywhere else
- squeaky voice in chrome indicates it is using the wrong sample rate on anything outside of firefox
- messing with the sample rate can lead to some funny outcomes, such as a very deep voice

**Comment by mmtmn** [comment]
if possible and if not fixed, please don't delete the branch, it might be helpful for future me

**Pull request closed without merging**